File: libs/moqtail-ts/src/model/control/server_setup.ts
import { ByteBuffer, FrozenByteBuffer, BaseByteBuffer } from '../common/byte_buffer'
import { ControlMessageType, DRAFT_11 } from './constant'
import { KeyValuePair } from '../common/pair'
import { LengthExceedsMaxError } from '../error/error'

export class ServerSetup {
  constructor(
    public readonly selectedVersion: number,
    public readonly setupParameters: KeyValuePair[],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.ServerSetup
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.ServerSetup)
    const payload = new ByteBuffer()
    payload.putVI(this.selectedVersion)
    payload.putVI(this.setupParameters.length)
    for (const param of this.setupParameters) {
      payload.putKeyValuePair(param)
    }
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('ServerSetup::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): ServerSetup {
    const selectedVersion = buf.getNumberVI()
    const paramCount = buf.getNumberVI()
    const setupParameters: KeyValuePair[] = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      setupParameters[i] = buf.getKeyValuePair()
    }
    return new ServerSetup(selectedVersion, setupParameters)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('ServerSetup', () => {
    test('roundtrip serialization/deserialization', () => {
      const selectedVersion = DRAFT_11
      const setupParameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Set me up!')),
      ]
      const msg = new ServerSetup(selectedVersion, setupParameters)
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.ServerSetup))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const parsed = ServerSetup.parsePayload(frozen)
      expect(parsed.selectedVersion).toBe(selectedVersion)
      expect(parsed.setupParameters).toEqual(setupParameters)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const selectedVersion = DRAFT_11
      const setupParameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Set me up!')),
      ]
      const msg = new ServerSetup(selectedVersion, setupParameters)
      const serialized = msg.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.ServerSetup))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const parsed = ServerSetup.parsePayload(frozen)
      expect(parsed.selectedVersion).toBe(selectedVersion)
      expect(parsed.setupParameters).toEqual(setupParameters)
      expect(frozen.remaining).toBe(3)
    })

    test('partial message fails', () => {
      const selectedVersion = DRAFT_11
      const setupParameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Set me up!')),
      ]
      const msg = new ServerSetup(selectedVersion, setupParameters)
      const serialized = msg.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        ServerSetup.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/server_setup.rs
use super::control_message::ControlMessageTrait;
use crate::model::common::pair::KeyValuePair;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{BufMut, Bytes, BytesMut};

use super::constant::ControlMessageType;

#[derive(Debug, PartialEq, Clone)]
pub struct ServerSetup {
  pub selected_version: u32,
  pub setup_parameters: Vec<KeyValuePair>,
}

impl ServerSetup {
  pub fn new(selected_version: u32, setup_parameters: Vec<KeyValuePair>) -> Self {
    Self {
      selected_version,
      setup_parameters,
    }
  }
}

impl ControlMessageTrait for ServerSetup {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::ServerSetup)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.selected_version)?;
    payload.put_vi(self.setup_parameters.len())?;

    for param in &self.setup_parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "ServerSetup::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);
    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let selected_version = payload.get_vi()? as u32;

    let number_of_parameters = payload.get_vi()?;
    let mut setup_parameters = Vec::new();

    for _ in 0..number_of_parameters {
      let param = KeyValuePair::deserialize(payload)?;
      setup_parameters.push(param);
    }

    Ok(Box::new(ServerSetup {
      selected_version,
      setup_parameters,
    }))
  }
  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::ServerSetup
  }
}

#[cfg(test)]
mod tests {
  use crate::model::control::constant::DRAFT_11;

  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let selected_version = DRAFT_11;
    let setup_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Set me up!")).unwrap(),
    ];
    let server_setup = ServerSetup {
      selected_version,
      setup_parameters,
    };

    let mut buf = server_setup.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::ServerSetup as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = ServerSetup::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, server_setup);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let selected_version = DRAFT_11;
    let setup_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Set me up!")).unwrap(),
    ];
    let server_setup = ServerSetup {
      selected_version,
      setup_parameters,
    };

    let serialized = server_setup.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::ServerSetup as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = ServerSetup::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, server_setup);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let selected_version = DRAFT_11;
    let setup_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Set me up!")).unwrap(),
    ];
    let server_setup = ServerSetup {
      selected_version,
      setup_parameters,
    };

    let mut buf = server_setup.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::ServerSetup as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = ServerSetup::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
